File: week14/comparative_language_analysis/game_of_life/lua/gol.lua
-------------------------------------------------------------------------------
-- Conway's game of life implemented in Lua. 
-- Simulation settings can be adjusted below in the global configurations

-- optional Args:
--      1. seedFile     - Path to a file containing a starting seed
--      2. height       - height of the simulation
--      3. width        - Width of the simulation
-------------------------------------------------------------------------------

-- Global Config
WIDTH = 24
HEIGHT = 24
FRAME_COUNT = 300
UPDATE_RATE = .1 -- Time between updates in seconds
DEFAULT_SEED = "../seeds/default.txt"

-- Tiles
ALIVE = "\u{2687}" -- ⚇
DEAD = "\u{0387}"  -- ·


-- Performance tracker
local Tracker = {}
Tracker.__index = Tracker

function Tracker.new(filename)
    local self = setmetatable({}, Tracker)
    self.frameCount = 0
    self.totalTime = 0
    self.totalMem = 0
    self.filename = filename

    -- Init log file
    local file, err = io.open(self.filename, "w")
    if file then
        file:write("Frame,CalcTime_s,MemUsage_KB,AvgTime_s,AvgMem_KB\n")
        file:close()
    else
        print("couldn't open file: " .. err)
        os.exit(1)
    end
    return self
end

function Tracker:update(frameTime, frameMem)
    self.frameCount = self.frameCount + 1
    self.totalTime = self.totalTime + frameTime
    self.totalMem = self.totalMem + frameMem
    
    local avgTime = self.totalTime / self.frameCount
    local avgMem = self.totalMem / self.frameCount
    
    -- Log to file
    local file = io.open(self.filename, "a")
    if file then
        file:write(string.format("%d,%8.6f,%8.4f,%8.6f,%8.4f\n", 
            self.frameCount, frameTime, frameMem, avgTime, avgMem))
        file:close()
    end
    
    return avgTime, avgMem
end


-- Grid class
local Grid = {}
Grid.__index = Grid

-- Constructor
function Grid.new(h, w)
    local self = setmetatable({}, Grid)

    -- Assign variables
    self.h = h
    self.w = w

    -- Build empty grid
    for y = 1, self.h do
        self[y] = {}
        for x = 1, self.w do
            self[y][x] = 0
        end
    end
    return self
end

-- clear the terminal and print the current grid state
function Grid:display(stats)
    -- Clear terminal using ANSI escape codes
    io.write("\27[H\27[2J")

    local output = ""
    for y = 1, self.h do
        for x = 1, self.w do
            output = output .. (self[y][x] == 1 and ALIVE or DEAD) .. " "
        end
        output = output .. "\n"
    end

    -- Append performance stats to output
    if stats then
        output = output .. "\n" .. string.rep("-", self.w * 2) .. "\n"
        output = output .. string.format("FRAME: %d\n", stats.count)
        output = output .. string.format("%-5s Current: %8.6f s  | Avg: %8.6f s\n", 
            "TIME:", stats.currTime, stats.avgTime)
        output = output .. string.format("%-5s Current: %8.4f KB | Avg: %8.4f KB\n", 
            "MEM:", stats.currMem, stats.avgMem)
    end
    print(output)
end

-- Seed
function Grid:seed(filename)
    local seedFile = io.open(filename, "r")
    if not seedFile then
        print("Could not open file: " .. filename)
        return false
    end

    -- Read file to get the width and height of the seed
    local lines = {}
    local seedWidth = 0
    for line in seedFile:lines() do
        table.insert(lines, line)
        if #line > seedWidth then
            seedWidth = #line
        end
    end
    seedFile:close()

    local seedHeight = #lines

    -- Calculate centered position
    local startY = math.floor((self.h - seedHeight) / 2) + 1
    local startX = math.floor((self.w - seedWidth) / 2) + 1

    for rowIdx, line in ipairs(lines) do
        local curRow = startY + rowIdx - 1
        -- Make sure we don't leave the grid
        if curRow >= 1 and curRow <= self.h then
            for i = 1, #line do
                local char = line:sub(i,i)
                local currentCol = startX + i - 1

                -- Make sure we don't leave the grid
                if currentCol >= 1 and currentCol <= self.w then
                    self[curRow][currentCol] = (char == "1") and 1 or 0
                end
            end
            curRow = curRow + 1
        end
    end
end

-- Count how many cells a neighbor has
function Grid:countNeighbors(y,x)
    local count = 0
    for dy = -1, 1 do
        for dx = -1, 1 do
            if not (dx == 0 and dy == 0) then
                local ny = (y + dy - 1) % self.h + 1
                local nx = (x + dx - 1) % self.w + 1
                count = count + self[ny][nx]
            end
        end
    end
    return count
end

-- Simulation Class
local Simulation = {}
Simulation.__index = Simulation

-- constructor
function Simulation.new(h,w)
    local self = setmetatable({}, Simulation)
    self.grid = Grid.new(h,w)
    return self
end

-- sleep function
function Simulation:sleep(n)
    local t0 = os.clock()
    while os.clock() - t0 <= n do
    end
end

function Simulation:update()
    local newGrid = Grid.new(self.grid.h, self.grid.w)

    for y = 1, self.grid.h do
        for x = 1, self.grid.w do
            local neighborCnt = self.grid:countNeighbors(y,x)
            local currentState = self.grid[y][x]

            if currentState == 1 then
                if neighborCnt < 2 or neighborCnt > 3 then
                    newGrid[y][x] = 0
                else
                    newGrid[y][x] = 1
                end
            else
                if neighborCnt == 3 then
                    newGrid[y][x] = 1
                else
                    newGrid[y][x] = 0
                end
            end
        end
    end
    self.grid = newGrid
end


local function main()
    -- init simulation
    local height = tonumber(arg[2]) or HEIGHT
    local width = tonumber(arg[3]) or WIDTH

    local simulation = Simulation.new(height, width)

    -- Load user seed
    local seed = arg[1] or DEFAULT_SEED
    simulation.grid:seed(seed)
    simulation.grid:display()

    -- Setup Tracker
    -- regex to get the seed name from the filename
    local baseName = seed:match("([^/\\]+)%.[^.]+$") or seed:match("([^/\\]+)$") or "seed"
    local outputFile = string.format("lua_%s.csv", baseName)
    local tracker = Tracker.new(outputFile)

    -- simulation loop
    for i = 1, FRAME_COUNT, 1 do
        local startTime = os.clock()
        simulation:update()
        local endTime = os.clock()
        local frameTime = endTime - startTime

        local frameMem = collectgarbage("count")

        local avgTime, avgMem = tracker:update(frameTime,frameMem)

        local stats = {
            count = i,
            currTime = frameTime,
            avgTime = avgTime,
            currMem = frameMem,
            avgMem = avgMem
        }

        simulation.grid:display(stats)

        simulation:sleep(UPDATE_RATE)
    end

end

main()

File: week14/comparative_language_analysis/game_of_life/julia/gol.jl
using Printf
using Statistics

###############################################################################
# Conway's game of life implemented in Lua. 
# Simulation settings can be adjusted below in the global configurations

# optional Args:
#      1. seedFile     - Path to a file containing a starting seed
#      2. height       - height of the simulation
#      3. width        - Width of the simulation
###############################################################################

# --- Global Config ---
const WIDTH = 24
const HEIGHT = 24
const FRAME_COUNT = 300
const UPDATE_RATE = 0.1 # Time between updates in seconds
const DEFAULT_SEED = "../seeds/default.txt"

# Tiles (Unicode strings)
const ALIVE = "\u2687" # ⚇
const DEAD  = "\u0387" # ·

# Performance Tracker 
mutable struct PerformanceTracker
    frame_count::Int
    total_time::Float64
    total_mem::Float64
    filename::String

    function PerformanceTracker(filename::String)
        # Initialize log file with header
        open(filename, "w") do io
            println(io, "Frame,CalcTime_s,MemUsage_KB,AvgTime_s,AvgMem_KB")
        end
        new(0, 0.0, 0.0, filename)
    end
end

function update_tracker!(tracker::PerformanceTracker, frame_time::Float64, frame_mem::Float64)
    tracker.frame_count += 1
    tracker.total_time += frame_time
    tracker.total_mem += frame_mem

    avg_time = tracker.total_time / tracker.frame_count
    avg_mem = tracker.total_mem / tracker.frame_count

    # Append to log file
    open(tracker.filename, "a") do io
        @printf(io, "%d,%8.6f,%8.2f,%8.6f,%8.2f\n", 
                tracker.frame_count, frame_time, frame_mem, avg_time, avg_mem)
    end

    return avg_time, avg_mem
end

# --- Grid Logic ---

function create_grid(w::Int, h::Int)
    return zeros(Int8, h, w) # Height is rows, Width is columns
end

function count_neighbors(grid::Matrix{Int8}, y::Int, x::Int)
    h, w = size(grid)
    count = 0
    for dy in -1:1, dx in -1:1
        (dy == 0 && dx == 0) && continue
        
        # Toroidal wrapping logic
        ny = mod1(y + dy, h)
        nx = mod1(x + dx, w)
        
        count += grid[ny, nx]
    end
    return count
end

function seed_grid!(grid::Matrix{Int8}, filename::String)
    if !isfile(filename)
        println("Could not open file: $filename")
        return false
    end

    lines = readlines(filename)
    seed_h = length(lines)
    seed_w = maximum(length, lines)

    grid_h, grid_w = size(grid)
    
    # Calculate centered position
    start_x = floor(Int, (grid_w - seed_w) / 2) + 1
    start_y = floor(Int, (grid_h - seed_h) / 2) + 1

    for (row_idx, line) in enumerate(lines)
        cur_y = start_y + row_idx - 1
        for (col_idx, char) in enumerate(line)
            cur_x = start_x + col_idx - 1
            if 1 <= cur_y <= grid_h && 1 <= cur_x <= grid_w
                grid[cur_y, cur_x] = (char == '1') ? 1 : 0
            end
        end
    end
end

function display_grid(grid::Matrix{Int8}, stats::Union{NamedTuple, Nothing}=nothing)
    # Clear terminal using ANSI escape codes
    print("\033[2J\033[H")
    
    h, w = size(grid)
    output = IOBuffer()

    for y in 1:h
        for x in 1:w
            print(output, grid[y, x] == 1 ? ALIVE : DEAD, " ")
        end
        print(output, "\n")
    end

    if stats !== nothing
        println(output, "\n", "-" ^ (w * 2))
        @printf(output, "FRAME: %d\n", stats.count)
        @printf(output, "%-5s Current: %8.6f s  | Avg: %8.6f s\n", "TIME:", stats.currTime, stats.avgTime)
        @printf(output, "%-5s Current: %8.2f KB | Avg: %8.2f KB\n", "MEM:", stats.currMem, stats.avgMem)
    end

    print(String(take!(output)))
end

# --- Simulation Logic (Functional Approach) ---

function update_grid(grid::Matrix{Int8})
    h, w = size(grid)
    
    # Using a functional map construction to create the new grid
    return [
        begin
            n = count_neighbors(grid, idx[1], idx[2])
            current = grid[idx]
            if current == 1
                (n == 2 || n == 3) ? Int8(1) : Int8(0)
            else
                (n == 3) ? Int8(1) : Int8(0)
            end
        end 
        for idx in CartesianIndices(grid)
    ]
end

# --- Main ---

function main()
    # Handle arguments
    seed_file = length(ARGS) >= 1 ? ARGS[1] : DEFAULT_SEED
    width = length(ARGS) >= 2 ? parse(Int, ARGS[2]) : WIDTH
    height = length(ARGS) >= 3 ? parse(Int, ARGS[3]) : HEIGHT

    grid = create_grid(width, height)
    seed_grid!(grid, seed_file)

    # Determine filename for CSV
    base_name = splitext(basename(seed_file))[1]
    output_file = "julia_$(base_name)_test.csv"
    tracker = PerformanceTracker(output_file)

    for i in 1:FRAME_COUNT
        # Timing and memory measurement
        # gc_live_bytes() provides a view of current heap usage
        start_time = time()
        
        # Generate the new grid
        grid = update_grid(grid)
        
        frame_time = time() - start_time
        frame_mem = Base.gc_live_bytes() / 1024.0

        avg_time, avg_mem = update_tracker!(tracker, frame_time, frame_mem)

        stats = (
            count = i,
            currTime = frame_time,
            avgTime = avg_time,
            currMem = frame_mem,
            avgMem = avg_mem
        )

        display_grid(grid, stats)
        sleep(UPDATE_RATE)
    end
end

# Run the program
main()